packages/core/src/events/embeddings/base.ts

```diff
@@ -1,36 +1,213 @@
-
-export abstract class Embeddings {
-    /**
-     * The async caller should be used by subclasses to make any async calls,
-     * which will thus benefit from the concurrency and retry logic.
-     */
-    caller: AsyncCaller;
-  
-    constructor(params: EmbeddingsParams) {
-      this.caller = new AsyncCaller(params ?? {});
+  AsyncCaller,
+  type AsyncCallerParams,
+} from '../../utils/asyncCaller.js';
+import { BaseEvent, BaseEventParams } from '../base.js';
+import { EmbedResult } from '../output/provide/embedresult.js';
+
+export interface BaseEmbeddingsCallOptions extends BaseEventParams {
+  /**
+   * Abort signal for the call.
+   * If provided, the call will be aborted when the signal is aborted.
+   */
+  signal?: AbortSignal;
+}
+
+export interface BaseEmbeddingsParams
+  extends AsyncCallerParams,
+    BaseEventParams {
+  cache?: BaseCache<number[]> | boolean;
+}
+
+export abstract class BaseEmbeddings<
+    CallOptions extends BaseEmbeddingsCallOptions = BaseEmbeddingsCallOptions,
+  >
+  extends BaseEvent<string, EmbedResult, CallOptions>
+  implements BaseEmbeddingsParams
+{
+  /**
+   * Represents the type for call options without some specified properties.
+   */
+  declare SerializedCallOptions: Omit<CallOptions, keyof CallableConfig>;
+
+  /**
+   * The async caller should be used by subclasses to make any async calls,
+   * which will thus benefit from the concurrency and retry logic.
+   */
+  caller: AsyncCaller;
+
+  /**
+   * Cache instance to store and retrieve results for given string.
+   */
+  cache?: BaseCache<number[]>;
+
+  _eventNamespace(): string[] {
+    return ['embeddings'];
+  }
+
+  constructor({ callbacks, ...params }: BaseEmbeddingsParams) {
+    super({ callbacks, ...params });
+
+    if (typeof params.cache === 'object') {
+      this.cache = params.cache;
+    } else if (params.cache) {
+      this.cache = MemoryCache.global<number[]>();
+    } else {
+      this.cache = undefined;
+    }
+
+    this.caller = new AsyncCaller(params ?? {});
+  }
+
+  /**
+   * Embeds a single document as input and returns a promise that resolves to a
+   * vector for the query document.
+   * @param input A single document to be embedded.
+   * @param options Additional options for embedding.
+   * @returns A promise that resolves to a vector for the query document.
+   */
+  async invoke(input: string, options?: CallOptions): Promise<EmbedResult> {
+    const result: EmbedResult = await this.embed(
+      input,
+      options,
+      options?.callbacks
+    );
+
+    return result;
+  }
+
+  /**
+   * Provides the core logic to interface with the language model, handling both cached and uncached embeddings.
+   * @param {string} document - A string.
+   * @param options - Optional call options.
+   * @param callbacks - Optional callbacks.
+   * @returns {Promise<EmbedResult>} The result from the language model embeddings.
+   */
+  async embed(
+    document: string,
```

Comment:
If i didn't remeber wrongly, document is originally of Document object/class. (Same with Chroma, and Langchain, not sure whether it would cause problems in the future if we change this data type to string here.

Reply:
hmmm, okay, I wonder whether that would cause any problem, say if our data/document is a JSON/CSV, where data has to be grouped together. But I guess for now if Langchain does it this way, then we can go with it.

Reply:
from langchain_community.document_loaders import JSONLoader
from langchain.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")

loader = JSONLoader(file_path="/content/data.json", jq_schema=".nodes[]", text_content=False)
documents = loader.load()

db = Chroma.from_documents(documents, embedding_function)

Reply:
This is doable in future. We just need a JSONLoader